### PR TITLE
fix: Favor `deref`fed types in "Generate function" assist

### LINKED ISSUE #9713: Generate function assist should favor &str over &String
Currently, if you use the generate function assist to turn produce a definition from an unresolved function call, the generated function will use the argument types verbatim from what you supplied. For some types this is almost never what you want, e.g. `&String` or `&Vec<_>`. Example:

```rust
let foo = String::new();
undefined_fn(&foo);

// generated:
fn undefined_fn(foo: &String) {
    todo!()
}
```

RA should choose deref-compatible types where appropriate.

### PR DESCRIPTION
**Pull request by mahdi-frms**
Fixes #9713

Although that's still not relying on sematic info.

### CODE REVIEW

File: crates/ide_assists/src/handlers/generate_getter.rs [hunk 1]
```diff
@@ -136,7 +136,7 @@ pub(crate) fn generate_getter_impl(
     )
 }
 
-fn useless_type_special_case(field_name: &str, field_ty: &ast::Type) -> Option<(String, String)> {
+pub fn useless_type_special_case(field_name: &str, field_ty: &String) -> Option<(String, String)> {
```
Comment (matklad) [inline crates/ide_assists/src/handlers/generate_getter.rs:139]: Let's pull this up into `assists/src/utils.rs` -- there shouldn't be uses "across" assists.

File: crates/ide_assists/src/handlers/generate_function.rs [hunk 2]
```diff
@@ -257,7 +258,19 @@ fn fn_args(
             None => String::from("arg"),
         });
         arg_types.push(match fn_arg_type(ctx, target_module, &arg) {
-            Some(ty) => ty,
+            Some(ty) => {
+                if ty.len() > 0 && ty.chars().next().unwrap() == '&' {
```
Comment (matklad) [inline crates/ide_assists/src/handlers/generate_function.rs:262]: ty.starts_with('&')

File: crates/ide_assists/src/handlers/generate_function.rs [hunk 3]
```diff
@@ -257,7 +258,19 @@ fn fn_args(
...
+                    if let Some((new_ty,_)) = useless_type_special_case("",&ty[1..].to_owned()){
+                        new_ty
+                    }
+                    else{
+                        ty
+                    }
```
Comment (matklad) [inline crates/ide_assists/src/handlers/generate_function.rs:268]: formatting is off here, I suggest configuring auto-format in your editor.

File: crates/ide_assists/src/handlers/generate_getter.rs [hunk 4]
```diff
@@ -154,7 +154,7 @@ fn useless_type_special_case(field_name: &str, field_ty: &ast::Type) -> Option<(
 }
 
 // FIXME: This should rely on semantic info.
-fn ty_ctor(ty: &ast::Type, ctor: &str) -> Option<String> {
+fn ty_ctor(ty: &String, ctor: &str) -> Option<String> {
```
Comment (matklad) [inline crates/ide_assists/src/handlers/generate_getter.rs:157]: It's unfortunate that we go from `ast` to `String` here, but it's also true that ast is not correct either -- we want `hir::ty` here.

**Review by matklad — COMMENTED** [review]
lgtm overall, added a couple of minor comments!

File: crates/ide_assists/src/handlers/generate_getter.rs (hunk 4 above)
Comment (mahdi-frms) [reply crates/ide_assists/src/handlers/generate_getter.rs:157]: OK, I don't have a clue how this should be done. Any suggestion?

Comment (matklad) [reply crates/ide_assists/src/handlers/generate_getter.rs:157]: That's mostly me thinking aloud -- I don't think there's an easy way to do that right now, so the current impl is fine!

That said, if you already can see a way to change both call-sites to pass in a [Type](https://github.com/rust-analyzer/rust-analyzer/blob/6aeaceaf45508383c4664bc253fb21e48034d1bc/crates/hir/src/lib.rs#L2137-L2142), than we should do that (in a separate PR). 


In other words, it's clear that we want this to operate on semantic notion of type (`hir::Type`). The way to do it though is to carefully peruse exiting APIs and usages and make that possible by extending the API. It's not just "use this magic function". 

In general, our overall semantic APIs are a mess at the moment, that's why we have  https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/Tree.20Based.20HIR discussion.

Comment (mahdi-frms) [reply crates/ide_assists/src/handlers/generate_getter.rs:157]: Well, currently I'm not that familiar with AST and HIR. Guess I'll leave it like this for now.

**Comment by mahdi-frms** [comment]
Needs the r+ by the way :)

**Comment by lnicola** [comment]
bors r=matklad

**Pull request merged**